source/game/kart/KartCollide.cc

```diff
@@ -0,0 +1,278 @@
+#include "KartCollide.hh"
+
+#include "game/kart/KartMove.hh"
+#include "game/kart/KartPhysics.hh"
+
+#include <game/field/CollisionDirector.hh>
+
+namespace Kart {
+
+KartCollide::KartCollide() : m_notTrickable(false) {}
+
+KartCollide::~KartCollide() = default;
+
+void KartCollide::init() {
+    m_50 = 0.0f;
+    m_offRoad = false;
+    m_groundBoostPanelOrRamp = false;
+    m_notTrickable = false;
+}
+
+void KartCollide::resetHitboxes() {
+    CollisionGroup *hitboxGroup = physics()->hitboxGroup();
+    for (u16 idx = 0; idx < hitboxGroup->hitboxCount(); ++idx) {
+        hitboxGroup->hitbox(idx).setLastPos(scale(), pose());
+    }
+}
+
+void KartCollide::updateHitboxes() {
+    CollisionGroup *hitboxGroup = physics()->hitboxGroup();
+    for (u16 idx = 0; idx < hitboxGroup->hitboxCount(); ++idx) {
+        hitboxGroup->hitbox(idx).update(move()->totalScale(), 0.0f, scale(), fullRot(), pos());
+    }
+}
+
+void KartCollide::findCollision() {
+    calcBodyCollision(move()->totalScale(), fullRot(), scale());
+}
+
+void KartCollide::calcBodyCollision(f32 totalScale, const EGG::Quatf &rot,
+        const EGG::Vector3f &scale) {
+    CollisionGroup *hitboxGroup = physics()->hitboxGroup();
+    CollisionData &collisionData = hitboxGroup->collisionData();
+    collisionData.reset();
+
+    EGG::Vector3f posRel;
+    Field::CourseColMgr::CollisionInfo colInfo;
+    colInfo.bbox.setDirect(EGG::Vector3f::zero, EGG::Vector3f::zero);
+    Field::KCLTypeMask maskOut;
+    EGG::BoundBox3f minMax = colInfo.bbox;
+
+    for (u16 hitboxIdx = 0; hitboxIdx < hitboxGroup->hitboxCount(); ++hitboxIdx) {
+        Field::KCLTypeMask flags = 0xEAFABDFF;
+        Hitbox &hitbox = hitboxGroup->hitbox(hitboxIdx);
+
+        if (hitbox.bspHitbox()->wallsOnly != 0) {
+            flags = 0x4A109000;
+        }
+
+        hitbox.update(totalScale, 0.0f, scale, rot, pos());
+
+        Field::CollisionDirector::Instance()->checkSphereCachedFullPush(hitbox.worldPos(),
+                hitbox.lastPos(), flags, &colInfo, &maskOut, hitbox.radius(), 0);
+    }
+
+    collisionData.speedFactor = 1.0f;
+    collisionData.rotFactor = 1.0f;
+}
+
+void KartCollide::calcFloorEffect() {
+    m_offRoad = true;
+    m_groundBoostPanelOrRamp = true;
+
+    Field::KCLTypeMask mask = KCL_NONE;
+    calcTriggers(&mask, pos(), false);
+
+    mask = KCL_NONE;
+    calcTriggers(&mask, pos(), true);
+}
+
+void KartCollide::calcTriggers(Field::KCLTypeMask *mask, const EGG::Vector3f &pos, bool twoPoint) {
+    EGG::Vector3f v1 = twoPoint ? physics()->pos() : EGG::Vector3f::inf;
+    Field::KCLTypeMask typeMask = twoPoint ? 0x05070000 : 0xE0F8BDFF;
```

Comment:
This should be rewritten to use the KCL `define`s. The first is `KCL_TYPE_DIRECTIONAL`. The second is `(KCL_TYPE_VEHICLE_COLLIDEABLE & ~KCL_TYPE_BIT(COL_TYPE_ITEM_STATE_MODIFIER) & ~KCL_TYPE_BIT(COL_TYPE_WEAK_WALL))` but I don't immediately know what to name it